transition.go

```diff
@@ -1,49 +1,120 @@
 package hsm
 
-// Transition represents a possible way of going from one state to another.
-// It contains the logic used to determine what the next step will be.
+import (
+	"reflect"
+)
+
+// Action is a type representing a function to be executed on a transition.
+type Action func(srv Service)
+
+// NopAction is a no-op action
+var NopAction = func(srv Service) {}
+
+// Transition defines the interface that is implemented by the different types of transitions
 type Transition interface {
-	NextState() State
+	Execute(srv Service, fromState State) State
+}
+
+// ExternalTransition is a transition from one state to a different state
+type ExternalTransition struct {
+	event   Event
+	toState State
+	action  Action
+}
+
+// ExternalTransition Constructor
+func NewExternalTransition(event Event, toState State, action Action) *ExternalTransition {
+	return &ExternalTransition{
+		event:   event,
+		toState: toState,
+		action:  action,
+	}
+}
+
+// Execute calls OnExit() for the fromState and all parent states up to the nil parent state
+// or the parentState that is shared by the new toState.  Action() is then called and
+// OnEnter() is finally called on the new toState.
+func (t *ExternalTransition) Execute(srv Service, fromState State) State {
+	// Call OnExit until one of the following is true:
+	//   - parentState is NilState
+	//   - parentState is equal to toState's parentState
+	parentState := fromState.OnExit(t.event)
+	for !IsNilState(parentState) {
```

Comment:
This looks like an unsafe infinite loop. For prod code, should there be a guard somewhere that prevents cycles in the hsm?

Reply:
The only way this can be infinite is if there is a circular state machine implementation.  However, that can happen so good catch.  I'll add code and a test to check and handle this condition.